src/testutils.rs

```diff
@@ -0,0 +1,124 @@
+use std::io;
+
+use async_trait::async_trait;
+use tokio::sync::{mpsc, watch};
+
+use crate::cid::{ConnectionId, ConnectionPeer};
+use crate::udp::AsyncUdpSocket;
+
+/// A mock socket that can be used to simulate a perfect link.
+#[derive(Debug)]
+pub struct MockUdpSocket {
+    outbound: mpsc::UnboundedSender<Vec<u8>>,
+    inbound: mpsc::UnboundedReceiver<Vec<u8>>,
+    /// Peers identified by a letter
+    pub only_peer: char,
+    /// Watch for signal that the link is up. If not up, link will SILENTLY drop all sent packets.
+    /// If up_status is None, the link is always up.
+    up_status: Option<watch::Receiver<bool>>,
+}
+
+impl MockUdpSocket {
+    /// Grab the status controller for this socket.
+    /// Can only be taken once, otherwise fails.
+    pub fn take_status_controller(&mut self) -> Result<watch::Sender<bool>, String> {
+        match self.up_status {
+            None => {
+                let (up_status_tx, up_status) = watch::channel(true);
+                self.up_status = Some(up_status);
+                Ok(up_status_tx)
+            }
+            Some(_) => Err("Status controller already taken".to_string()),
+        }
+    }
+
+    /// Should the link send packets?
+    fn active(&self) -> bool {
+        match &self.up_status {
+            None => true,
+            Some(up_status) => *up_status.borrow(),
+        }
+    }
+}
+
+#[async_trait]
+impl AsyncUdpSocket<char> for MockUdpSocket {
+    async fn send_to(&mut self, buf: &[u8], target: &char) -> io::Result<usize> {
+        if target != &self.only_peer {
+            panic!("MockUdpSocket only supports sending to one peer");
+        }
+        if !self.active() {
+            return Ok(buf.len());
+        }
+        let result = self.outbound.send(buf.to_vec());
+        if let Err(err) = result {
+            Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                format!("channel closed: {err}"),
+            ))
+        } else {
+            Ok(buf.len())
+        }
+    }
+
+    async fn recv_from(&mut self, buf: &mut [u8]) -> io::Result<(usize, char)> {
+        let packet = self
+            .inbound
+            .recv()
+            .await
+            .ok_or_else(|| io::Error::new(io::ErrorKind::UnexpectedEof, "channel closed"))?;
+        if buf.len() < packet.len() {
+            panic!("buffer too small for perfect link");
+        }
+        let packet_len = packet.len();
+        buf[..packet_len].copy_from_slice(&packet[..]);
+        Ok((packet_len, self.only_peer))
+    }
+}
+
+impl ConnectionPeer for char {}
+
+pub fn build_link_pair() -> (MockUdpSocket, MockUdpSocket) {
+    let (peer_a, peer_b): (char, char) = ('A', 'B');
```

Comment:
> nit: I would prefer something more in line with the `(tx, rx) = channel()` semantics

Hm, are you talking about the refactor I mentioned in #100 ?

>>  one natural path is to split the socket into a sender and receiver (much how creating a channel returns a split tx and rx object)

That refactor is interesting, but feels significant and out of scope. I'm not really sure what it would look like without changing how the socket trait is defined.

> and ideally the connection ID stuff would also be taken care of here, since there is only ever one connection per socket. if that can't be done easily here, then we can come back to it, but I think it would be nice to unify the two steps to enforce consistency.

You're thinking like a single method that calls both `build_link_pair()` and `build_connection_id_pair()` and then returns the 4 resulting values?

Reply:
the last bit is what I'm referring to yes, returning a pair of pairs. and then those other functions don't need to be `pub`.